transport/src/main/java/io/netty/channel/AbstractCoalescingBufferQueue.java

```diff
@@ -0,0 +1,268 @@
+                bytes -= entryBuffer.readableBytes();
+            }
+        }
+        readableBytes -= originalBytes - bytes;
+        assert readableBytes >= 0;
+        return toReturn;
+    }
+
+    /**
+     * The number of readable bytes.
+     */
+    public int readableBytes() {
+        return readableBytes;
+    }
+
+    /**
+     * Are there pending buffers in the queue.
+     */
+    public boolean isEmpty() {
+        return bufAndListenerPairs.isEmpty();
+    }
+
+    /**
+     *  Release all buffers in the queue and complete all listeners and promises.
+     */
+    public void releaseAndFailAll(ChannelOutboundInvoker invoker, Throwable cause) {
+        releaseAndCompleteAll(invoker.newFailedFuture(cause));
+    }
+
+    /**
+     * Copy all pending entries in this queue into the destination queue.
+     * @param dest to copy pending buffers to.
+     */
+    public void copyTo(AbstractCoalescingBufferQueue dest) {
+        dest.bufAndListenerPairs.addAll(bufAndListenerPairs);
+        dest.readableBytes += readableBytes;
+    }
+
+    /**
+     * Writes all remaining elements in this queue.
+     * @param ctx The context to write all elements to.
+     */
+    public void writeAndRemoveAll(ChannelHandlerContext ctx) {
+        readableBytes = 0;
+        Throwable pending = null;
+        ByteBuf previousBuf = null;
+        for (;;) {
+            Object entry = bufAndListenerPairs.poll();
+            if (entry == null) {
+                if (previousBuf != null) {
+                    ctx.write(previousBuf, ctx.voidPromise());
+                }
+                break;
+            }
+            try {
+                if (entry instanceof ByteBuf) {
+                    if (previousBuf != null) {
+                        ctx.write(previousBuf, ctx.voidPromise());
+                    }
+                    previousBuf = (ByteBuf) entry;
+                } else if (entry instanceof ChannelPromise) {
+                    ctx.write(previousBuf, (ChannelPromise) entry);
+                    previousBuf = null;
+                } else {
+                    ctx.write(previousBuf).addListener((ChannelFutureListener) entry);
+                    previousBuf = null;
+                }
+            } catch (Throwable t) {
+                if (pending == null) {
+                    pending = t;
+                }
+            }
+        }
+        if (pending != null) {
+            throw new IllegalStateException(pending);
+        }
+    }
+
+    /**
+     * Calculate the result of {@code current + next}.
+     */
+    protected abstract ByteBuf compose(ByteBufAllocator alloc, ByteBuf cumulation, ByteBuf next);
+
+    /**
+     * Calculate the first {@link ByteBuf} which will be used in subsequent calls to
+     * {@link #compose(ByteBufAllocator, ByteBuf, ByteBuf)}.
+     */
+    protected ByteBuf composeFirst(ByteBufAllocator allocator, ByteBuf first) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused method parameter "allocator". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)